react-renderer/src/App.tsx

```diff
@@ -51,12 +51,27 @@ class App extends React.Component<any, ICanvasState> {
   public readonly canvas = React.createRef<Canvas>();
   public readonly buttons = React.createRef<ButtonBar>();
 
+  public modShapes = async (state: State) => {
+    this.modCanvas(state); // is this the right way to call it
+  };
+
   public onConnectionStatus = (conn: ConnectionStatus) => {
     Log.info(`Connection status: ${conn}`);
   };
   public onVersion = (version: string) => {
     this.setState({ penroseVersion: version });
   };
+  // same as onCanvasState but doesn't alter timeline or involve optimization
+  // used only in modshapes
+  public modCanvas = async (canvasState: State) => {
+    await new Promise((r) => setTimeout(r, 1));
+
+    await this.setState({
+      data: canvasState,
+      history: [...this.state.history],
```

Comment:
you can actually omit `history: [...this.state.history],` altogether -- I recently learned that React merges state updates https://reactjs.org/docs/state-and-lifecycle.html#state-updates-are-merged